language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,164 @@
+pub fn validate_signed_transaction<A: VMAdapter>(
+    adapter: &A,
+    transaction: SignedTransaction,
+    state_view: &dyn StateView,
+) -> VMValidatorResult {
+    let _timer = TXN_VALIDATION_SECONDS.start_timer();
+    let txn_sender = transaction.sender();
+    let log_context = AdapterLogSchema::new(state_view.id(), 0);
+
+    let txn = if let Ok(t) = A::check_signature(transaction) {
+        t
+    } else {
+        return VMValidatorResult::error(StatusCode::INVALID_SIGNATURE);
+    };
+
+    let remote_cache = StateViewCache::new(state_view);
+    let account_role = get_account_role(txn_sender, &remote_cache);
+    let mut session = adapter.new_session(&remote_cache);
+
+    let (status, normalized_gas_price) = match validate_signature_checked_transaction(
+        adapter,
+        &mut session,
+        &txn,
+        &remote_cache,
+        true,
+        &log_context,
+    ) {
+        Ok((price, _)) => (None, price),
+        Err(err) => (Some(err.status_code()), 0),
+    };
+
+    // Increment the counter for transactions verified.
+    let counter_label = match status {
+        None => "success",
+        Some(_) => "failure",
+    };
+    TRANSACTIONS_VALIDATED
+        .with_label_values(&[counter_label])
+        .inc();
+
+    VMValidatorResult::new(status, normalized_gas_price, account_role)
+}
+
+fn get_account_role(sender: AccountAddress, remote_cache: &StateViewCache) -> GovernanceRole {
+    let role_access_path = create_access_path(sender, RoleId::struct_tag());
+    if let Ok(Some(blob)) = remote_cache.get(&role_access_path) {
+        return bcs::from_bytes::<account_config::RoleId>(&blob)
+            .map(|role_id| GovernanceRole::from_role_id(role_id.role_id()))
+            .unwrap_or(GovernanceRole::NonGovernanceRole);
+    }
+    GovernanceRole::NonGovernanceRole
+}
+
+pub(crate) fn validate_signature_checked_transaction<S: MoveStorage, A: VMAdapter>(
+    adapter: &A,
+    mut session: &mut Session<S>,
+    transaction: &SignatureCheckedTransaction,
+    remote_cache: &S,
+    allow_too_new: bool,
+    log_context: &AdapterLogSchema,
+) -> Result<(u64, Identifier), VMStatus> {
+    if transaction.is_multi_agent() && !adapter.allows_multi_agent()? {
+        // Multi agent is not allowed
+        return Err(VMStatus::Error(StatusCode::FEATURE_UNDER_GATING));
+    }
+
+    if transaction.contains_duplicate_signers() {
+        return Err(VMStatus::Error(StatusCode::SIGNERS_CONTAIN_DUPLICATES));
+    }
+
+    let gas_price = transaction.gas_unit_price();
+    let currency_code_string = transaction.gas_currency_code();
+    let currency_code = match account_config::from_currency_code_string(currency_code_string) {
+        Ok(code) => code,
+        Err(_) => {
+            return Err(VMStatus::Error(StatusCode::INVALID_GAS_SPECIFIER));
+        }
+    };
+
+    let normalized_gas_price = match get_currency_info(&currency_code, remote_cache) {
+        Ok(info) => info.convert_to_xdx(gas_price),
```

Comment:
This is pretty Diem-specific, but getting rid of it seems hard... 

I need to think more, but I think an ideal adapter trait would assume that there's only one gas currency and that all prices are normalized to this currency. Individual implementations can support paying for gas in multiple currencies, but they will need to implement the normalization on their own.

Reply:
Do we need paying for gas a part of the trait at all?

payment feels like an application specific thing, while gas is a more general concept.

Reply:
Good question--I'm not totally sure. I wouldn't want the concept of paying for gas in the VM, but it might be appropriate  to have it in the adapter trait. It's a fairly common practice, and the mempool needs to (or at least can benefit from) knowing about gas payments to filter txes that will fail because an account has spent all its funds on gas.

Reply:
I included `get_gas_price` in the trait, nothing about the payment part though. I think gas is a common enough blockchain concept for it to be mentioned in the trait? We can revisit this decision when the mempool interface is made more generic. Right now we have to mention gas and governance roles because mempool expects these pieces of information to rank the txns.